packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+        trimmedSigners[i] = signers[i];
+      }
+      return (trimmedAccounts, trimmedIssuedOns, trimmedSigners);
+    } else {
+      return (accounts, issuedOns, signers);
+    }
+  }
+
+  /**
+   * @notice Returns identifiers mapped (by unrevoked signers) to a
+   *   given account address by a list of trusted issuers
+   * @param account Address of the account
+   * @param trustedIssuers Array of n issuers whose identifier mappings will be used
+   * @param maxIdentifiers Limit the number of identifiers that will be returned
+   * @return Array (length <= maxIdentifiers) of identifiers
+   * @dev Adds identifier info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+
+  // TODO ASv2 consider also returning an array with counts of identifiers per issuer
+
+  function lookupIdentifiersByAddress(
+    address account,
+    address[] memory trustedIssuers,
+    uint256 maxIdentifiers
+  ) public view returns (bytes32[] memory) {
+    // Same as for the other lookup, preallocate and then trim for now
+    uint256 currIndex = 0;
+    bytes32[] memory identifiers = new bytes32[](maxIdentifiers);
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      address trustedIssuer = trustedIssuers[i];
+      uint256 numIssuersForAddress = addressToIdentifiers[account][trustedIssuer].length;
+      for (uint256 j = 0; j < numIssuersForAddress; j = j.add(1)) {
+        // Iterate through the list of identifiers
+        if (currIndex < maxIdentifiers) {
+          bytes32 identifier = addressToIdentifiers[account][trustedIssuer][j];
+          // Check if the mapping was produced by a revoked signer
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation[] memory attestationsForIssuer = identifierToAddresses[identifier][trustedIssuer];
+          for (uint256 k = 0; k < attestationsForIssuer.length; k = k.add(1)) {
+            OwnershipAttestation memory attestation = attestationsForIssuer[k];
+            // (identifier, account, issuer) tuples should be unique
+            if (attestation.account == account && !revokedSigners[attestation.signer]) {
+              identifiers[currIndex] = identifier;
+              currIndex = currIndex.add(1);
+              break;
+            }
+          }
+        } else {
+          break;
+        }
+      }
+    }
+    if (currIndex < maxIdentifiers) {
+      // Allocate and fill properly-sized array
+      bytes32[] memory trimmedIdentifiers = new bytes32[](currIndex);
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedIdentifiers[i] = identifiers[i];
+      }
+      return trimmedIdentifiers;
+    } else {
+      return identifiers;
+    }
+  }
+
+  // TODO do we want to restrict permissions, or should anyone
+  // with a valid signature be able to register an attestation?
+  modifier isValidUser(address issuer, address account) {
```

Comment:
Mixing checks for "is this sender the user" and "is this sender the issuer" in the same modifier seems odd to me. Not what I expected from the name of the function.

Also nit: `isValidUser` is a boolean "question". Modifier are access controllers and so should be something more like `onlyValidUser`.

Reply:
Hm, "user" her is meant to refer to whoever is interacting with the contract, I guess `onlyValidSender` would be a better name?